### PR TITLE
Add support for GPU inference

### PR DESCRIPTION
**Pull request by sh7dm**
Address #688 .

Tasks:
- [x] `GPUTensor`
  - [x] Import
  - [x] Export
  - [x] Intermediate data in GPU memory
- [ ] Ops
  - [ ] Convolution
  - [ ] Activations
    - [x] tanh
    - [x] sigmoid
    - [ ] relu
  - [ ] Fully-connected
  - [ ] Pooling
  - [ ] Softmax
- [ ] Runner for models
  - [ ] Managing GPU memory
    - [ ] Free buffers no longer in use to allow for models larger than GPU memory
- [ ] Examples working
  - [ ] `tensorflow-mobilenet-v2`

### CODE REVIEW

File: gpu/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+repository = "https://github.com/snipsco/tract"
+keywords = ["TensorFlow", "NeuralNetworks"]
+categories = ["science"]
+autobenches = false
+edition = "2018"
+
+[badges]
+maintenance = { status = "actively-developed" }
+
+[dependencies]
```
Comment (sh7dm) [inline gpu/Cargo.toml:16]: Are to be decluttered later, some might not be necessary

**Comment by kali** [comment]
This is a good start. 
* I'm a bit curious about what are you plans for the operator list ? Do you plan on supporting everything in tract-core ? Or just this subset ?
* Also, wondering about he "right" way to implement sigmoid and tanh on gpu. You used the exponentiation based definition, but I found out that using a rationale development was faster on cpu. See https://github.com/sonos/tract/blob/main/linalg/src/generic/sigmoid.rs for instance.

**Comment by sh7dm** [comment]
Strides patch is on its way making tanh over 1000x1000 1.5x more efficient on RADV than on llvmpipe.

1. Everything I guess, implementing one-by-one with some goals like getting MobileNet running.
2. Hmm, they're builtin. I didn't care a lot about best algos on GPU, top priority for now is managing tensors out there. Then optimizations might take place.

**Comment by sh7dm** [comment]
Also wanted to ask you: do we really have 4-dimensional tensors? If not, will it be fine to crash when one encountered? If so, how to handle them gracefully when wgpu only has 3D workgroup queuing?

**Comment by kali** [comment]
4D dimensional tensors are the norm, not the exception. All image categorisers (inception, mobilenet, all of them really) work in NCHW or NHWC. Most frameworks will support up to 8 dimensions.

**Comment by sh7dm** [comment]
Hmm, well, if I think about it then it's obvious: n feature maps each with 3 color planes each with x and y. How to get it into 3D properly? Split the tensor?

**Comment by sh7dm** [comment]
Well, a solution might be that a uniform introduced and we run multiple workgroups with different values of the parameter to work on different 3D objects from 4D list of 3D objects

**Comment by sh7dm** [comment]
@haixuanTao could you please give a piece of advice or some idea for fixing that issue?

**Comment by haixuanTao** [comment]
> This is a good start.
> 
>     * I'm a bit curious about what are you plans for the operator list ? Do you plan on supporting everything in tract-core ? Or just this subset ?
> 
>     * Also, wondering about he "right" way to implement sigmoid and tanh on gpu. You used the exponentiation based definition, but I found out that using a rationale development was faster on cpu. See https://github.com/sonos/tract/blob/main/linalg/src/generic/sigmoid.rs for instance.

Fyi, exponential precision is pretty bad: https://github.com/webonnx/wonnx/issues/87#issuecomment-1086916156

**Comment by sh7dm** [comment]
Okay, thank you, I'll take action on that later, after basic principles

**Comment by haixuanTao** [comment]
> @haixuanTao could you please give a piece of advice or some idea for fixing that issue?

In wonnx, we only use 1-D array representing the concatenation of all dimensions and reference as C-like arrays.

**Comment by sh7dm** [comment]
Well, that looks like z = z*w, will do that and invent other fix if we run into z*w > 65535.

**Comment by sh7dm** [comment]
The problem with this way of handling is passing W coordinate into the shader. X, Y and Z are managed by workgroup creation, but to acquire W dividing Z by W size will be necessary.

**Comment by haixuanTao** [comment]
> Well, that looks like z = z×w, will do that and invent other fix if we run into z×w > 65535. Thanks for info

There is many optimization like vectorisation that we have already implemented in wonnx. I genuinely feel like you're just redoing many of the things, we already did in wonnx. 

> The problem with this way of handling is passing W coordinate into the shader. X, Y and Z are managed by workgroup creation, but to acquire W dividing Z by W size will be necessary.

In wonnx, we have all the information required when creating workgroup, so it shouldn't be a problem.

I genuinely feel like you're going to reimplement many things we already did in wonnx, and I feel like, as a community, we should probably try to find a way to merge those two projects, avoid having two different implementations, find a compromise, as the maintenance burden is quite high, especially as `wgpu` is still not that stable.

But, as I'm not part of the `Tract` project, and therefore, this decision is not mine to make.

**Comment by sh7dm** [comment]
Hmm, I'm not either, but open to collaboration.

**Comment by sh7dm** [comment]
About integrating wonnx into Tract (from my point of view): implementation of stuff like tensors should be consistent with Tract (what I'm working on now in this PR), but stuff like shader template compilation might be reviewed and integrated.

**Comment by kali** [comment]
Well, in terms of architecture, I think the best way to go is to make a bridge between tract-core and wgpu, using work done for wonnx. I am ok with light alterations to tract to make it easier, like making public stuff which is not yet in tract-core.

For a deeper integration in tract, I will only consider it with 1/ a (quasi-)total coverage of tract-core features and operators 2/ some form of maintainer commitment as I can not maintain the gpu stuff myself on top of the rest. So it's okay to "experiment" in tract repository, obviously, it gives us a convenient discussion medium, but I will not merge short of these two constraints.

**Comment by sh7dm** [comment]
> Well, in terms of architecture, I think the best way to go is to make a bridge between tract-core and wgpu, using work done for wonnx. I am ok with light alterations to tract to make it easier, like making public stuff which is not yet in tract-core.
> 
> For a deeper integration in tract, I will only consider it with 1/ a (quasi-)total coverage of tract-core features and operators 2/ some form of maintainer commitment as I can not maintain the gpu stuff myself on top of the rest. So it's okay to "experiment" in tract repository, obviously, it gives us a convenient discussion medium, but I will not merge short of these two constraints.

This. I think compiler from wonnx can be used with tract-core. Coverage should be pretty much okay. Maintenance? Well, either me (not sure if able to respond quickly enough) or @haixuanTao, if you're willing to do so.
@haixuanTao now it's your time to think about that idea, are you willing to integrate some works of wonnx into Tract and maintain it. Not sure if integration of the two projects will open up all the greatest features of both, but that looks achievable.

**Comment by sh7dm** [comment]
In this plan we do some model runner which runs Tract's `SimplePlan` over my `GPUTensor`s with help of wonnx's shader compiler. Seems workable.

**Comment by haixuanTao** [comment]
For me, `wonnx` is still not at a (quasi-)total coverage of tract-core features and operators. And so, integration to tract at the moment seems not desirable. 

I'm ok to maintain it. What I would propose is that, `wonnx` becomes the `Tract-GPU` extension per se ( name might or might not change ). But this will require effort from wonnx to integrate tract logic. This will not add too much load on Tract side.

**Comment by sh7dm** [comment]
Yes, seems like the solution does not need a lot, if any, support from tract-core. Most stuff is already exposed, you get a decluttered plan and do accelerated ops on it.

**Comment by kali** [comment]
> For me, `wonnx` is still not at a (quasi-)total coverage of tract-core features and operators. And so, integration to tract at the moment seems not desirable.

Agreed. So this bridge should probably start its existence outside of tract, as a separate entity. Just prove the concept, see if it gets traction, if it becomes mature, we can discuss integration.
 
> I'm ok to maintain it. What I would propose is that, `wonnx` becomes the `Tract-GPU` extension per se ( name might or might not change ). But this will require effort from wonnx to integrate tract logic. This will not add too much load on Tract side.

Agreed, I think most of the work happens around or within wonnx. I don't mind keeping the discussion happening here if it helps, but at this stage, that's about it.

**Comment by kali** [comment]
By the way folks, I spell `tract` with a lowercase initial ;)

**Comment by haixuanTao** [comment]
Yes, I think it would be the best place to have it maintained and we would be able to run CI on it as well.

**Comment by sh7dm** [comment]
Should I fork wonnx tomorrow and move all the code I have to there?

**Comment by haixuanTao** [comment]
I think that a PR would be great yeah! It would be awesome :)

**Comment by kali** [comment]
Since you've found a new home, I'm going to close this issue ;)
